### ISSUE TITLE
kernel_constraint = constraint_nonneg() not functioning correctly / negative outputs

### ISSUE THREAD

**Issue by atroiano** [rstudio/keras#418]
I have a model that only has non negative values in all the inputs and targets.  The output layer keeps determining negative numbers.  My data is currently normalized but has some pretty heavy skewing.  There are lots of instances where the target variable is 0.

The target variable is revenue so it should never be negative.  I tried to add kernel_constraint = constraint_nonneg() to the final layer and it does not work.   Below is my code.  The output is the next three months revenue.

```embedding_hospital_out <- input_hospital %>% 
  layer_embedding(
    input_dim = embed_hospital[[2]]$num_words+1
    , output_dim = embed_hospital_size
    , name="embedding_hospital") %>%  
  layer_dropout(rate = drop_out_rate) %>% 
  layer_flatten()

combined_model <- layer_concatenate(c(embedding_hospital_out,input_dense)) %>%
  layer_dense(units=units_1
              ,kernel_initializer = 'glorot_normal'
              #,activation='relu'
              # ,activity_regularizer =regularizer_l1_l2(l1 = 0.01, l2 = 0.01)
              ,name = 'layer_1'
              
  ) %>%
  layer_activation_leaky_relu(alpha = .3) %>% 
  layer_batch_normalization() %>%
  layer_dropout(rate = drop_out_rate) %>% 
  layer_dense(units=units_2
              ,kernel_initializer = 'glorot_normal'
             # ,activation='relu'
              #,activity_regularizer =regularizer_l1_l2(l1 = 0.01, l2 = 0.01)
              ,name = 'layer_2'
  ) %>%
  layer_activation_leaky_relu(alpha = .3) %>% 
  layer_batch_normalization() %>%
  layer_dropout(rate = drop_out_rate) %>%
  layer_dense(units=units_2
              ,kernel_initializer = 'glorot_normal'
             # ,activation='relu'
              ,name = 'layer_3'
              #,activity_regularizer =regularizer_l1_l2(l1 = 0.01, l2 = 0.01)
  ) %>%
 layer_activation_leaky_relu(alpha = .3) %>% 
  layer_batch_normalization() %>%
  layer_dropout(rate = drop_out_rate) %>% 
  layer_gaussian_noise(stddev = 1) %>%
  layer_dense(
    units = target_dim
    ,kernel_constraint = constraint_nonneg()
    ,name = 'layer_output'
  ) ```

**Comment by MaximilianPi** [comment]
> The output layer keeps determining negative numbers. 

By using leaky relu before your last layer, the activity can be negative before your last layer and the kernel_constraint is only applied to the weights, so your last layer even requires the ability to learn negative weights to switch the final output back to positive.  (Of course this scenario requires the sum(neg) > sum(pos) at output node_i )
Therefore you should use a relu layer after the third layer (before the final layer) + kernel_constraint or use leaky_ relu without nonneg constraint.

I think the problem lies elsewhere:
Can you please tell something about your data dimensions and the network architecture? I think your network has maybe way too many parameters and you cannot find an optimum.

**Comment by atroiano** [comment]
The data for this is pretty small, ~9k features rows with ~15 numeric features in the dense input and  my embedding layer is 8 words and 1100 unique number of words.  

Here is my model architecture:

```
__________________________________________________________________________________________________________________________________
Layer (type)                              Output Shape                 Param #         Connected to                               
==================================================================================================================================
input_hosptial (InputLayer)               (None, 8)                    0                                                          
__________________________________________________________________________________________________________________________________
embedding_hospital (Embedding)            (None, 8, 2)                 2228            input_hosptial[0][0]                       
__________________________________________________________________________________________________________________________________
dropout_1 (Dropout)                       (None, 8, 2)                 0               embedding_hospital[0][0]                   
__________________________________________________________________________________________________________________________________
flatten_1 (Flatten)                       (None, 16)                   0               dropout_1[0][0]                            
__________________________________________________________________________________________________________________________________
input_1 (InputLayer)                      (None, 15)                   0                                                          
__________________________________________________________________________________________________________________________________
concatenate_1 (Concatenate)               (None, 31)                   0               flatten_1[0][0]                            
                                                                                       input_1[0][0]                              
__________________________________________________________________________________________________________________________________
layer_1 (Dense)                           (None, 500)                  16000           concatenate_1[0][0]                        
__________________________________________________________________________________________________________________________________
leaky_re_lu_1 (LeakyReLU)                 (None, 500)                  0               layer_1[0][0]                              
__________________________________________________________________________________________________________________________________
batch_normalization_1 (BatchNormalization (None, 500)                  2000            leaky_re_lu_1[0][0]                        
__________________________________________________________________________________________________________________________________
dropout_2 (Dropout)                       (None, 500)                  0               batch_normalization_1[0][0]                
__________________________________________________________________________________________________________________________________
layer_2 (Dense)                           (None, 150)                  75150           dropout_2[0][0]                            
__________________________________________________________________________________________________________________________________
leaky_re_lu_2 (LeakyReLU)                 (None, 150)                  0               layer_2[0][0]                              
__________________________________________________________________________________________________________________________________
batch_normalization_2 (BatchNormalization (None, 150)                  600             leaky_re_lu_2[0][0]                        
__________________________________________________________________________________________________________________________________
dropout_3 (Dropout)                       (None, 150)                  0               batch_normalization_2[0][0]                
__________________________________________________________________________________________________________________________________
layer_output (Dense)                      (None, 3)                    453             dropout_3[0][0]                            
==================================================================================================================================
Total params: 96,431
Trainable params: 95,131
Non-trainable params: 1,300
__________________________________________________________________________________________________________________________________

````

**Comment by MaximilianPi** [comment]
So your target variables range from 0 to 300k? Maybe you should transform it with log(y+1). I'm not so sure why are you using three outputs? Which loss function are you using?

**Comment by atroiano** [comment]
my outputs are the next three months of revenue, so for 2017-01-01, my targets are  2017-02-01, 2017-03-01, 2017-04-01 revenue.  I am using MSE

**Comment by MaximilianPi** [comment]
Have you tried it with log(y+1) transformation and a smaller architecture?

**Comment by atroiano** [comment]
Yeah, the two changes that helped the most were removing the leaky relu from the first layer, log transforming all my input and targets values and making the network smaller.

I still get negative values but those values are are tiny.
